src/tool_provider/artifactory.rs

```diff
@@ -18,19 +26,120 @@ impl ArtifactoryProvider {
         Self { paths }
     }
 }
-#[allow(unused)]
+
 impl ToolProviderImpl for ArtifactoryProvider {
-    fn get_releases(&self, repo: &str) -> ForemanResult<Vec<Release>> {
-        Err(ForemanError::Other {
-            message: "Artifactory is not yet supported. Please use Github or Gitlab as your source"
-                .to_owned(),
-        })
+    fn get_releases(&self, repo: &str, host: &Url) -> ForemanResult<Vec<Release>> {
+        let client = Client::new();
+
+        let url = format!("{}artifactory/api/storage/{}", host, repo);
+        let params = vec![("list", ""), ("deep", "1")];
+        let mut builder = client
+            .get(&url)
+            .header(USER_AGENT, "Roblox/foreman")
+            .query(&params);
+
+        let tokens = artiaa_auth::Tokens::load(&self.paths.artiaa_path()?)
+            .map_err(|error| ForemanError::ArtiAAError { error })?;
+
+        if let Some(credentials) = tokens.get_credentials(host) {
+            builder = builder.header(ARTIFACTORY_API_KEY_HEADER, credentials.token.to_string());
+        }
+        log::debug!("Downloading artifactory releases for {}", repo);
+        let response_body = builder
+            .send()
+            .map_err(ForemanError::request_failed)?
+            .text()
+            .map_err(ForemanError::request_failed)?;
+
+        let response: ArtifactoryResponse =
+            serde_json::from_str(&response_body).map_err(|err| {
+                ForemanError::unexpected_response_body(err.to_string(), response_body, url)
+            })?;
+
+        let mut releases: Vec<ArtifactoryRelease> = Vec::new();
+        let mut release_map: HashMap<&str, Vec<ArtifactoryAsset>> = HashMap::new();
+        for file in &response.files {
+            let mut uri = file.uri.split("/");
+            // file.uri will look something like /<version>/<artifact-name>, so uri will be ["", <version>, <artifact-name]
+            uri.next();
+            let version = uri.next().unwrap();
+            let asset_name = uri.next().unwrap();
+
+            let asset_url = format!("{}artifactory/{}/{}/{}", host, repo, version, asset_name);
+
+            let asset = ArtifactoryAsset {
+                url: asset_url,
+                name: asset_name.to_string(),
+            };
+
+            release_map.entry(version).or_insert(Vec::new()).push(asset);
+        }
+
+        for (version, assets) in release_map {
+            releases.push(ArtifactoryRelease {
+                tag_name: version.to_string(),
+                assets,
+            })
+        }
```

Comment:
Not a big deal, but I think this could be expressed a little more elegantly with a functional approach.